### ISSUE TITLE
Request: Configurable priority between misaligned accesses and general access traps

### ISSUE THREAD

**Issue by billmcspadden-riscv** [riscv/sail-riscv#192]
Reference:  Priv Spec., Table 3.7 "Synchronous exception priority in decreasing priority order."

The priority of a "load/store/AMO address misaligned" exception can be made optionally
higher or lower than other memory access exceptions.  The model needs to be made
configurable to handle this option.

This option will need to be added to the RISCV-Config structure.

The solution to this problem should use the "new" YAML configuration methods that have been introduced to the
model (see PR #175).

**Comment by Timmmm** [comment]
This is now supported. The high priority check (before address translation) is controlled by `memory.misaligned.supported` in the config, and the low priority check (after address translation) is controlled by `memory.regions[N].misaligned_fault`.

Hello btw, I hope retirement is treating you well!

**Comment by allenjbaum** [comment]
I don't think I understand how this maps to implementing the optional
behaviour.

Does.  memory.regions[N].misaligned_fault  indicate that, when there is
both a misaligned access and a page fault, that the misaligned access gets
reported?
And, does memory.misaligned.supported that indicate that any non-naturally
aligned access causes a misaligned exception?
I would expect implementations to have misaligned granularities, and those
granularities would be different in different regions (e.g. cacheline in
memory, naturally aligned or 4B/8B granularity in IO regions),
and allowed to be different in different regions.

I don't see how you can get that behavior with a scalar configuration
parameter and a region based vector configuration parameters

**Comment by Timmmm** [comment]
It does this:

1. Before translation, if the access is misaligned and `memory.misaligned.supported` is `false` then you get an exception (high priority). Otherwise...
2. Perform address translation.
3. Look up the PMA for the physical address.
4. If the address is misaligned and `memory.regions[N].misaligned_fault` is not `"None"` then you get an exception (low priority) - either misaligned or access fault depending on what `memory.regions[N].misaligned_fault` is set to.

> I would expect implementations to have misaligned granularities, and those granularities would be different in different regions (e.g. cacheline in memory, naturally aligned or 4B/8B granularity in IO regions), and allowed to be different in different regions.

If the above checks pass, then there's a separate setting that controls how the misaligned memory access is decomposed into memory operations:

* `memory.misaligned.allowed_within_exp` - this controls the size of regions for which the memory access can still be done as a single misaligned operation, instead of splitting it into multiple aligned operations. If it is 0 then it's always split into aligned operations. 
* `memory.misaligned.byte_by_byte` - if this is `true` then misaligned accesses that are split into parts will be split into individual bytes, otherwise they'll be split into the minimum number aligned operations possible.
* `memory.misaligned.order_decreasing` - this just controls the order of operations.

We probably have some more options to add, e.g:

1. The `memory.misaligned` options could be moved to be PMA parameters (the reason they aren't is that they were added before PMAs were supported). 
2. There's no way to control the relative order of address translation and memory operations for split accesses (i.e. do all translations first, or do translation, operation, translation, operation).

But we do cover both exception priorities at least.

Also this could be better documented!

**Comment by allenjbaum** [comment]
The thing that jumped out at me was step 1: what exactly your definition of
"address is misaligned".
Based on the way this is defined, I would guess that if
memory.misaligned.supported  is false, it means accesses are never
decomposed,
but it also implies that misaligned must have higher priority than page
fault.
And, granularity could be defined to be >4K, so you always
determine misalignment from a VA - you need to translate first.

The spec does (as I read it) allow sa misaligned access to be signaled as a
page fault if there is one.
I'm pretty sure that isn't the intent, but that is how it is written. It
would be lovely to clarify...

**Comment by Timmmm** [comment]
> step 1: what exactly your definition of "address is misaligned".

If it's not naturally aligned.

> I would guess that if memory.misaligned.supported  is false, it means accesses are never decomposed,

Correct.

> but it also implies that misaligned must have higher priority than page fault.

Also correct.

> And, granularity could be defined to be >4K, so you always determine misalignment from a VA - you need to translate first.

Not sure I follow this...

> The spec does (as I read it) allow sa misaligned access to be signaled as a page fault if there is one. I'm pretty sure that isn't the intent, but that is how it is written. It would be lovely to clarify...

Yeah I definitely don't think that's what they were going for!

**Comment by allenjbaum** [comment]
So, step 1 is a special case for naturally aligned granularity. But, the
way the spec was written, you could still page-fault in that case.
It is a stupid implementation on a number of fronts (and the fact that the
spec allows it isn't so wonderful either)

But for >4K granularity, something that is in the middle of an 64K
superpage could have VA translations that span a much larger range,
so it can't choose which exception to signal until after translation.

**Comment by Timmmm** [comment]
Yes the misaligned atomicity granule PMA is for *physical* addresses, so it has to be handled after address translation in every case irrespective of the granule size.

> you could still page-fault in that case

Ah you mean you could check the natural alignment of the virtual address after doing address translation (and page faulting)? Yeah I agree the spec is written ambiguously to sort-of allow that. It would definitely be a very weird implementation though so I don't think we should support it (and you can get the same behaviour just by allowing misaligned virtual addresses but then disallowing misaligned physical addresses in all your PMAs).

**Comment by allenjbaum** [comment]
Hmm - not sure what it means to "allow" misaligned virtual addresses. PTEs
don't do it, only PMAs do it, so the "misaligned supported"=1 when no PMA
supports it sounds like an illegal configuration.
"misaligned supported" isn't architectural, its a construct we are using to
parameterize the claimed behavior of implementation.  It needs to be
consistent.
The optional behavior of which cuase has precedence is architectural, but
underdefined (IMHO).
That said: having Sail do it this way is OK, until somebody does it
differently and claims spec conformance, and they'd be right. I think we
both judge it would be unlikely.
